### PR TITLE
[InstCombine] Scalarize `(vec_ops (insert ?, X, Idx))` when only one element is demanded

### PR DESCRIPTION
**Pull request by goldsteinn**
This came as a result of PR #84389 (see: https://github.com/dtcxzyw/llvm-opt-benchmark/pull/336#discussion_r1518183485). SLP vectorizer can vectorize in a
pattern like:
```
    (blend
        (vec_ops0... (insert ?,X,0)),
        (vec_ops1... (insert ?,Y,1))
    )
```

In this case, `vec_ops0...` and `vec_ops1...` are essentially doing
scalar transforms.

We previously we handle things like:
`(blend (insert ?,X,0), (insert ?,Y,0))`

This patch extends that to look through `vec_ops...` that can be
scalarized, and if its possible to scalarize all ops, it transforms
the input to:
```
    (blend
        (insert ?,(scalar_ops0... X), 0),
        (insert ?,(scalar_ops1... Y), 0)
    )
```

### CODE REVIEW

File: llvm/test/Transforms/InstCombine/insert-extract-shuffle-inseltpoison.ll [hunk 1]
```diff
@@ -547,8 +547,7 @@ define <4 x float> @insert_in_splat_variable_index(float %x, i32 %y) {
 
 define <4 x float> @insert_in_nonsplat(float %x, <4 x float> %y) {
 ; CHECK-LABEL: @insert_in_nonsplat(
-; CHECK-NEXT:    [[XV:%.*]] = insertelement <4 x float> poison, float [[X:%.*]], i64 0
-; CHECK-NEXT:    [[SPLAT:%.*]] = shufflevector <4 x float> [[XV]], <4 x float> [[Y:%.*]], <4 x i32> <i32 poison, i32 0, i32 4, i32 poison>
```
Comment (goldsteinn) [inline llvm/test/Transforms/InstCombine/insert-extract-shuffle-inseltpoison.ll:551]: This is buggy, fixing.

**Comment by goldsteinn** [comment]
Note, the new function `getScalarizationOfInsertElement` is quite similiar to `cheapToScalarize`. I couldn't really see a way to properly merge the two, but that might be a better direction.

**Comment by goldsteinn** [comment]
@dtcxzyw can you test this. IMO its not worth the extra complexity unless it shows up in realworld.

**Comment by RKSimon** [comment]
Are you saying that SLP is creating these vectorizations but shouldn't? Isnt that a cost issue with the alt opcode cases?

**Comment by dtcxzyw** [comment]
> Are you saying that SLP is creating these vectorizations but shouldn't? Isnt that a cost issue with the alt opcode cases?

See the test `replace_through_casts` in `llvm/test/Transforms/InstCombine/shufflevector-inselt.ll`. The second element of `ui_v` and the first element of `si_v` are unused. Is this behavior expected in SLP?

**Comment by alexey-bataev** [comment]
> Are you saying that SLP is creating these vectorizations but shouldn't? Isnt that a cost issue with the alt opcode cases?

Does not look like cost issue, looks like a corner case that should be handled separately in SLP. We have some kind of it already, but not all. There is something similar with sext/zext/trunc, I'm going to fix later, since it requires some extra stuff in minbitwidth analysis. This one should not require extra work, just need to automatically create gather node for 2-element altshuffle, if we see that the operand is going to be gather/buildvector node. Can do it tomorrow.

**Comment by goldsteinn** [comment]
> > Are you saying that SLP is creating these vectorizations but shouldn't? Isnt that a cost issue with the alt opcode cases?
> 
> Does not look like cost issue, looks like a corner case that should be handled separately in SLP. We have some kind of it already, but not all. There is something similar with sext/zext/trunc, I'm going to fix later, since it requires some extra stuff in minbitwidth analysis. This one should not require extra work, just need to automatically create gather node for 2-element altshuffle, if we see that the operand is going to be gather/buildvector node. Can do it tomorrow.

Thanks :)

Ill close this once your patch goes up.

**Comment by dtcxzyw** [comment]
Ping @alexey-bataev

Any progress?

**Comment by alexey-bataev** [comment]
https://github.com/llvm/llvm-project/pull/84978

**Pull request closed without merging**